### PR TITLE
Avoid jobs without work load in join hash

### PR DESCRIPTION
**Pull request by toniSta**
Currently, there are many jobs created during the Hash Join without any work load. In my setup (benchmark with scale factor 0.01, 100 runs) almost 55 % of the jobs scheduled during the build phase had nothing to do. The graphic below shows the distribution of partition sizes within this phase. A partition size ("value") of 0 means, the job has no workload.

screenshot from 2018-07-03 17-25-09

### CODE REVIEW

File: src/lib/operators/join_hash.cpp [hunk 1]
```diff
@@ -129,16 +129,17 @@ std::vector<std::shared_ptr<HashTable<HashedType>>> build(const RadixContainer<L
...
-      // Prune empty partitions, so that we don't have too many empty hash tables
-      if (partition_size == 0) {
-        return;
-      }
+    // Prune empty partitions, so that we don't have too many empty hash tables
+    if (partition_size == 0) {
+      continue;
+    }
+
+    jobs.emplace_back(std::make_shared<JobTask>([&, current_partition_id, partition_size]() {
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: This looks like you are capturing the extracted variables by reference. By the time the lambda is run, things like `partition_left_begin` might have already changed. Have you tested this in multi-threaded mode?

File: src/lib/operators/join_hash.cpp [hunk 2]
```diff
@@ -375,17 +376,17 @@ void probe(const RadixContainer<RightType>& radix_container,
...
+
+    // Skip empty partitions to avoid empty output chunks
+    if ((partition_end - partition_begin) == 0) {
+      continue;
+    }
+
     jobs.emplace_back(std::make_shared<JobTask>([&, current_partition_id]() {
       // Get information from work queue
       auto& partition = static_cast<Partition<RightType>&>(*radix_container.elements);
-      const auto& partition_begin = radix_container.partition_offsets[current_partition_id];
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: same here

File: src/lib/operators/join_hash.cpp [hunk 3]
```diff
@@ -454,16 +455,17 @@ void probe_semi_anti(const RadixContainer<RightType>& radix_container,
...
```
Comment (mrks) [inline src/lib/operators/join_hash.cpp]: and again

File: src/lib/operators/join_hash.cpp [hunk 4]
```diff
@@ -375,17 +376,17 @@ void probe(const RadixContainer<RightType>& radix_container,
 
   for (size_t current_partition_id = 0; current_partition_id < (radix_container.partition_offsets.size() - 1);
        ++current_partition_id) {
+    const auto& partition_begin = radix_container.partition_offsets[current_partition_id];
+    const auto& partition_end = radix_container.partition_offsets[current_partition_id + 1];
+
+    // Skip empty partitions to avoid empty output chunks
+    if ((partition_end - partition_begin) == 0) {
```
Comment (lawben) [inline src/lib/operators/join_hash.cpp]: isn't this just a complicated way of writing: `partition_end == partition_begin`?

File: src/lib/operators/join_hash.cpp [hunk 5]
```diff
@@ -454,16 +455,17 @@ void probe_semi_anti(const RadixContainer<RightType>& radix_container,
...
```
Comment (lawben) [inline src/lib/operators/join_hash.cpp]: same as above

File: src/lib/operators/join_hash.cpp (hunk 1 above)
Comment (toniSta) [reply src/lib/operators/join_hash.cpp]: I have tested it in multi-threaded mode. The things haven't changed when the lambda was run.
However, passing the [u]ints by reference doesn't make sense.

Comment (mrks) [reply src/lib/operators/join_hash.cpp]: Well no - you *should not* pass them by reference, but by copy. By adding a sleep and then printing `partition_left_begin`, you should see that they are all referring to the same value.

Comment (toniSta) [reply src/lib/operators/join_hash.cpp]: Okay. I am passing the variables by copy now.

**Comment by Bouncner** [comment]
Has anybody looked into the sort-merge join?

**Pull request merged**